### PR TITLE
An example of BFS bottom-up approach using pykokkos

### PR DESCRIPTION
**Pull request by Wilson1211**
Given an n x m binary matrix, we want to return the distance of the nearest 0 for each cell as well as the distance farthest from cell 0. 
The distance between two adjacent cells is 1.
For example, given matrix:
0,0,0
0,1,0
1,1,1
we need to show that the distance of each cell to 0 is  
0,0,0
0,1,0
1,2,1
Also, the program will show the farthest distance to cell 0, in this example, the farthest distance to cell 0 is 2.

### CODE REVIEW

File: examples/kokkos-tutorials/workload/bfs_bottomup.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+
+@pk.workload
+class Workload:
+    def __init__(self, N: int, M: int):
+        self.N: int = N
+        self.M: int = M
+
+        self.val: pk.View1D[pk.double] = pk.View([N*M], pk.double)
+        self.vis: pk.View1D[pk.double] = pk.View([N*M], pk.double)
+        self.Matrix: pk.View2D[pk.double] = pk.View([N, M], pk.double)
```
Comment (gliga) [inline examples/kokkos-tutorials/workload/bfs_bottomup.py:15]: Matrix -> mat

File: examples/kokkos-tutorials/workload/bfs_bottomup.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+        self.max_arr: pk.View1D[pk.double] = pk.View([N], pk.double)
+        self.max_arr2D: pk.View2D[pk.double] = pk.View([N, N], pk.double)
+
+        self.timer_result: float = 0
+        self.element: float = N*M
+
+
```
Comment (gliga) [inline examples/kokkos-tutorials/workload/bfs_bottomup.py:22]: remove an extra empty line

File: examples/kokkos-tutorials/workload/bfs_bottomup.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+        for i in range(N*M):
+            self.val[i] = N+M
+
+        for i in range(M):
+            self.vis[i] = 0
+
```
Comment (gliga) [inline examples/kokkos-tutorials/workload/bfs_bottomup.py:28]: remove an extra empty line

File: examples/kokkos-tutorials/workload/bfs_bottomup.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+    def run(self):
+        timer = pk.Timer()
+        self.timer_result = timer.seconds()
+
+        # do the bfs
+        for i in range(self.N+self.M):
+            pk.parallel_for(self.element, self.check_vis)
+
+        # after bfs find maximum value in each row
+        pk.parallel_for("02", self.N, self.findmax)
```
Comment (gliga) [inline examples/kokkos-tutorials/workload/bfs_bottomup.py:49]: for the first argument in this call and calls below, you want to use/give unique names

File: examples/kokkos-tutorials/workload/bfs_bottomup.py [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+from parse_args import parse_args
+
+@pk.workload
+class Workload:
+    def __init__(self, N: int, M: int):
+        self.N: int = N
+        self.M: int = M
+
+        self.val: pk.View1D[pk.double] = pk.View([N*M], pk.double)
+        self.vis: pk.View1D[pk.double] = pk.View([N*M], pk.double)
```
Comment (gliga) [inline examples/kokkos-tutorials/workload/bfs_bottomup.py:14]: is vis for visited? if yes, I would suggest expanding the name and using pk.int type

File: examples/kokkos-tutorials/workload/bfs_bottomup.py [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+        var_col: int = i % self.M
+        min_val: float = self.val[i]
+
+        flag: int = 0
+
+        if (self.Matrix[var_row][var_col]==0) and (self.vis[i] == 0):
+            self.vis[i] = 1
+            self.val[i] = 0
+        else:
+            if i>=self.M:
```
Comment (gliga) [inline examples/kokkos-tutorials/workload/bfs_bottomup.py:84]: a comment for each branch would help follow code in this block

**Review by gliga — COMMENTED** [review]
mostly minor comments from my end.

I wish there was a way to provide an input from outside the workload and then we test it with various inputs, but this is fine for the first iteration

File: examples/kokkos-tutorials/workload/bfs_bottomup.py [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+        self.Matrix: pk.View2D[pk.double] = pk.View([N, M], pk.double)
+        self.max_arr: pk.View1D[pk.double] = pk.View([N], pk.double)
+        self.max_arr2D: pk.View2D[pk.double] = pk.View([N, N], pk.double)
+
+        self.timer_result: float = 0
+        self.element: float = N*M
+
+
+        for i in range(N*M):
+            self.val[i] = N+M
```
Comment (NaderAlAwar) [inline examples/kokkos-tutorials/workload/bfs_bottomup.py:24]: Consider using the view `fill()` method to initialize views. You can apply this to all views in this class.

File: examples/kokkos-tutorials/workload/bfs_bottomup.py [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+################################
+    @pk.workunit
+    def check_vis(self, i:int):
+        var_row: int = i//self.M
+        var_col: int = i % self.M
+        min_val: float = self.val[i]
+
+        flag: int = 0
+
+        if (self.Matrix[var_row][var_col]==0) and (self.vis[i] == 0):
```
Comment (NaderAlAwar) [inline examples/kokkos-tutorials/workload/bfs_bottomup.py:80]: No need for parentheses here

**Review by NaderAlAwar — COMMENTED** [review]
Looks good overall. I will take another look once the comments are addressed.

File: examples/kokkos-tutorials/workload/bfs_bottomup.py [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+class Workload:
+    def __init__(self, N: int, M: int):
+        self.N: int = N
+        self.M: int = M
+
+        self.val: pk.View1D[pk.double]       = pk.View([N*M], pk.double)
+        self.visited: pk.View1D[int]         = pk.View([N*M], int)
+        self.mat: pk.View2D[pk.double]       = pk.View([N, M], pk.double)
+        self.max_arr: pk.View1D[pk.double]   = pk.View([N], pk.double)
+        self.max_arr2D: pk.View2D[pk.double] = pk.View([N, N], pk.double)
```
Comment (NaderAlAwar) [inline examples/kokkos-tutorials/workload/bfs_bottomup.py:17]: Try to follow the formatting in the other examples, no need for this many spaces

File: examples/kokkos-tutorials/workload/bfs_bottomup.py [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+        # Initialize the input matrix, can be design to be any binary matrix
+        # In this example, mat[0][1] & mat[0][3] will be 0, others will be 1
+        self.mat.fill(1)
+        self.mat[0][1] = 0
+        self.mat[0][3] = 0
+
+    @pk.main
+    def run(self):
+        timer = pk.Timer()
+        self.timer_result = timer.seconds()
```
Comment (NaderAlAwar) [inline examples/kokkos-tutorials/workload/bfs_bottomup.py:35]: At the moment, this timer isn't really timing anything. You can either remove it or move the call to `timer.seconds()`.

**Review by NaderAlAwar — CHANGES_REQUESTED** [review]

**Comment by NaderAlAwar** [comment]
I would also move the location from `examples/kokkos-tutorials/workload` to `examples/graphs` (create a new directory)

**Review by NaderAlAwar — CHANGES_REQUESTED** [review]
The `parse_args.py` file is not needed here. You should instead use the `argparse` module directly in `bfs_bottump.py`, and you should only include the arguments you need, which are `M`, `N`, and `space` here

**Review by NaderAlAwar — APPROVED** [review]
Looks good, thanks!

**Pull request merged**